File: adws/adw_triggers/adw_trigger_cron_teamwork_tasks.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = [
#   "pydantic",
#   "python-dotenv",
#   "click",
#   "rich",
#   "schedule",
# ]
# ///
"""
Teamwork Task Monitor - Continuous polling for agent-ready tasks.

Monitors a Teamwork project for tasks with execution triggers and delegates them
to appropriate workflow scripts. Runs continuously with configurable polling interval.
"""

import os
import sys
import time
import json
import logging
import argparse
import subprocess
from pathlib import Path
from typing import List, Optional
from datetime import datetime

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from adw_modules.data_models import TeamworkTask, TeamworkCronConfig
from adw_modules.agent import execute_template, AgentTemplateRequest, generate_short_id
from adw_modules.utils import parse_json

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


class TeamworkTaskManager:
    """Manages fetching and delegating tasks from Teamwork."""

    def __init__(self, config: TeamworkCronConfig):
        self.config = config
        self.project_id = config.project_id
        self.status_mapping = config.status_mapping
        self.active_adw_ids = set()

    def get_eligible_tasks(self, limit: int = 10) -> List[TeamworkTask]:
        """Fetch eligible tasks from Teamwork project."""
        logger.info(f"Fetching tasks from Teamwork project {self.project_id}")

        # Build request for /get_teamwork_tasks
        request = AgentTemplateRequest(
            agent_name="teamwork-task-fetcher",
            slash_command="/get_teamwork_tasks",
            args=[
                self.project_id,
                json.dumps(self.config.status_filter),
                str(limit)
            ],
            adw_id=generate_short_id(),
            model="sonnet",
            working_dir=os.getcwd()
        )

        # Execute via agent framework
        try:
            response = execute_template(request)

            if not response.success:
                logger.error(f"Failed to fetch Teamwork tasks: {response.output}")
                return []

            # Parse JSON response
            task_data = parse_json(response.output, list)
            if not task_data:
                logger.info("No tasks returned from Teamwork")
                return []

            # Convert to TeamworkTask objects
            tasks = []
            for task_dict in task_data:
                try:
                    teamwork_task = TeamworkTask(
                        task_id=task_dict.get("task_id"),
                        project_id=task_dict.get("project_id"),
                        title=task_dict.get("title", ""),
                        status=task_dict.get("status"),
                        description=task_dict.get("description", ""),
                        tags=task_dict.get("tags", {}),
                        execution_trigger=task_dict.get("execution_trigger"),
                        task_prompt=task_dict.get("task_prompt", ""),
                        worktree=task_dict.get("tags", {}).get("worktree"),
                        model=task_dict.get("tags", {}).get("model"),
                        workflow_type=task_dict.get("tags", {}).get("workflow"),
                        prototype=task_dict.get("tags", {}).get("prototype"),
                        assigned_to=task_dict.get("assigned_to"),
                        created_time=task_dict.get("created_time"),
                        due_date=task_dict.get("due_date")
                    )

                    if teamwork_task.is_eligible_for_processing():
                        tasks.append(teamwork_task)
                        logger.info(f"Found eligible task: {teamwork_task.task_id} - {teamwork_task.title}")
                    else:
                        logger.debug(f"Skipping ineligible task: {teamwork_task.task_id}")

                except Exception as e:
                    logger.error(f"Failed to parse task: {e}")
                    continue

            return tasks

        except Exception as e:
            logger.error(f"Exception while fetching tasks: {e}", exc_info=True)
            return []

    def update_task_status(
        self,
        task_id: str,
        status: str,
        update_content: str = ""
    ) -> bool:
        """Update Teamwork task status and post comment."""
        logger.info(f"Updating task {task_id} to status: {status}")

        # Map system status to Teamwork status
        teamwork_status = self.config.map_status_to_teamwork(status)

        # Build update request
        request = AgentTemplateRequest(
            agent_name="teamwork-task-updater",
            slash_command="/update_teamwork_task",
            args=[
                task_id,
                teamwork_status,
                update_content
            ],
            adw_id=generate_short_id(),
            model="sonnet",
            working_dir=os.getcwd()
        )

        # Execute update
        try:
            response = execute_template(request)

            if response.success:
                logger.info(f"Successfully updated task {task_id} to {teamwork_status}")
                return True
            else:
                logger.error(f"Failed to update task {task_id}: {response.output}")
                return False

        except Exception as e:
            logger.error(f"Exception while updating task {task_id}: {e}")
            return False

    def delegate_task(self, task: TeamworkTask) -> bool:
        """Delegate task to appropriate workflow script."""
        # Generate ADW ID
        adw_id = generate_short_id()

        # Check for duplicates
        if adw_id in self.active_adw_ids:
            logger.warning(f"Duplicate ADW ID detected: {adw_id}, regenerating...")
            adw_id = generate_short_id()

        self.active_adw_ids.add(adw_id)

        # Get model preference
        model = task.get_preferred_model()

        # Get worktree name
        worktree_name = task.worktree or task.tags.get("worktree")
        if not worktree_name:
            # Generate from task description
            worktree_name = self._generate_worktree_name(task)

        # Get task prompt
        task_prompt = task.get_task_prompt_for_agent()

        logger.info(f"Delegating task {task.task_id} with ADW ID {adw_id}")
        logger.info(f"  Model: {model}")
        logger.info(f"  Worktree: {worktree_name}")
        logger.info(f"  Workflow: {self._determine_workflow(task)}")

        # Claim task immediately
        update_metadata = {
            "adw_id": adw_id,
            "timestamp": datetime.now().isoformat(),
            "model": model,
            "worktree_name": worktree_name,
            "status": "In progress"
        }

        if not self.update_task_status(
            task.task_id,
            "In progress",
            json.dumps(update_metadata)
        ):
            logger.error(f"Failed to claim task {task.task_id}")
            self.active_adw_ids.discard(adw_id)
            return False

        # Determine workflow script
        if task.prototype:
            script_path = "./adws/adw_plan_implement_update_teamwork_task.py"
            script_args = [
                adw_id,
                task.task_id,
                task_prompt,
                worktree_name,
                task.prototype,
                model,
                self.project_id
            ]
        elif task.should_use_full_workflow():
            script_path = "./adws/adw_plan_implement_update_teamwork_task.py"
            script_args = [
                adw_id,
                task.task_id,
                task_prompt,
                worktree_name,
                "",  # No prototype
                model,
                self.project_id
            ]
        else:
            script_path = "./adws/adw_build_update_teamwork_task.py"
            script_args = [
                adw_id,
                task.task_id,
                task_prompt,
                worktree_name,
                model,
                self.project_id
            ]

        if self.config.dry_run:
            logger.info(f"[DRY RUN] Would spawn: {script_path} {' '.join(script_args)}")
            return True

        # Spawn detached subprocess
        try:
            subprocess.Popen(
                [script_path] + script_args,
                start_new_session=True,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
            logger.info(f"Successfully spawned workflow for task {task.task_id}")
            return True

        except Exception as e:
            logger.error(f"Failed to spawn workflow: {e}")
            # Update task to Failed status
            error_metadata = {
                "adw_id": adw_id,
                "error": str(e),
                "timestamp": datetime.now().isoformat()
            }
            self.update_task_status(
                task.task_id,
                "Failed",
                json.dumps(error_metadata)
            )
            self.active_adw_ids.discard(adw_id)
            return False

    def _determine_workflow(self, task: TeamworkTask) -> str:
        """Determine which workflow to use for task."""
        if task.prototype:
            return f"plan-implement ({task.prototype})"
        elif task.should_use_full_workflow():
            return "plan-implement"
        else:
            return "build"

    def _generate_worktree_name(self, task: TeamworkTask) -> str:
        """Generate a worktree name from task title."""
        import re
        # Simple sanitization
        name = task.title.lower()
        name = re.sub(r'[^a-z0-9]+', '-', name)
        name = name.strip('-')
        # Limit length
        if len(name) > 50:
            name = name[:50]
        return f"proto-{name}" if task.prototype else f"feat-{name}"

    def run_once(self) -> int:
        """Run a single polling cycle. Returns number of tasks processed."""
        logger.info("=== Starting polling cycle ===")

        # Get eligible tasks
        tasks = self.get_eligible_tasks(limit=self.config.max_concurrent_tasks)

        if not tasks:
            logger.info("No eligible tasks found")
            return 0

        logger.info(f"Found {len(tasks)} eligible task(s)")

        # Delegate each task
        processed = 0
        for task in tasks:
            if self.delegate_task(task):
                processed += 1
            time.sleep(1)  # Brief delay between delegations

        logger.info(f"=== Polling cycle complete: {processed}/{len(tasks)} tasks delegated ===")
        return processed

    def run_continuous(self):
        """Run continuous polling loop."""
        logger.info("Starting Teamwork task monitor")
        logger.info(f"Project ID: {self.project_id}")
        logger.info(f"Polling interval: {self.config.polling_interval}s")
        logger.info(f"Max concurrent tasks: {self.config.max_concurrent_tasks}")
        logger.info(f"Status filter: {self.config.status_filter}")
        logger.info(f"Dry run: {self.config.dry_run}")

        cycle_count = 0

        try:
            while True:
                cycle_count += 1
                logger.info(f"\n--- Cycle {cycle_count} ---")

                self.run_once()

                logger.info(f"Sleeping for {self.config.polling_interval}s...")
                time.sleep(self.config.polling_interval)

        except KeyboardInterrupt:
            logger.info("\nReceived interrupt signal, shutting down gracefully...")
        except Exception as e:
            logger.error(f"Fatal error in polling loop: {e}")
            raise


def main():
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description="Teamwork task monitor for multi-agent prototyping system"
    )
    parser.add_argument(
        "--once",
        action="store_true",
        help="Run once and exit (no continuous monitoring)"
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Run in dry-run mode without making changes"
    )
    parser.add_argument(
        "--interval",
        type=int,
        default=15,
        help="Polling interval in seconds (default: 15)"
    )
    parser.add_argument(
        "--max-tasks",
        type=int,
        default=3,
        help="Maximum concurrent tasks (default: 3)"
    )

    args = parser.parse_args()

    # Load configuration from environment
    project_id = os.getenv("TEAMWORK_PROJECT_ID")
    if not project_id:
        logger.error("TEAMWORK_PROJECT_ID environment variable is required")
        sys.exit(1)

    # Build configuration
    config = TeamworkCronConfig(
        project_id=project_id,
        polling_interval=args.interval,
        max_concurrent_tasks=args.max_tasks,
        dry_run=args.dry_run
    )

    # Create manager
    manager = TeamworkTaskManager(config)

    # Run
    if args.once:
        logger.info("Running in single-cycle mode")
        processed = manager.run_once()
        logger.info(f"Processed {processed} task(s)")
        sys.exit(0)
    else:
        manager.run_continuous()


if __name__ == "__main__":
    main()

File: adws-bun/src/triggers/adw-trigger-cron-teamwork-tasks.ts
#!/usr/bin/env bun
/**
 * Teamwork Task Monitor - Continuous polling for agent-ready tasks.
 *
 * Monitors a Teamwork project for tasks with execution triggers and delegates them
 * to appropriate workflow scripts. Runs continuously with configurable polling interval.
 *
 * Usage:
 *   bun run src/triggers/adw-trigger-cron-teamwork-tasks.ts [options]
 *   ./src/triggers/adw-trigger-cron-teamwork-tasks.ts [options]
 *
 * Options:
 *   --once           Run once and exit (no continuous monitoring)
 *   --dry-run        Run in dry-run mode without making changes
 *   --interval N     Polling interval in seconds (default: 15)
 *   --max-tasks N    Maximum concurrent tasks (default: 3)
 */

import { parseArgs } from 'util';
import { z } from 'zod';

// Import core modules
import { executeTemplate, generateShortId, getSafeSubprocessEnv } from '@/modules/agent';
import {
  type TeamworkTask,
  type TeamworkCronConfig,
  type AgentTemplateRequest,
  TeamworkTaskSchema,
  TeamworkCronConfigSchema,
  isTeamworkTaskEligibleForProcessing,
  getTeamworkPreferredModel,
  getTeamworkTaskPromptForAgent,
  shouldTeamworkTaskUseFullWorkflow,
  mapStatusToTeamwork,
} from '@/modules/data-models';
import { parseJson } from '@/modules/utils';

// ============================================================================
// Logger
// ============================================================================

const logger = {
  info: (...args: unknown[]) => {
    console.log(`[${new Date().toISOString()}] [INFO]`, ...args);
  },
  warn: (...args: unknown[]) => {
    console.warn(`[${new Date().toISOString()}] [WARN]`, ...args);
  },
  error: (...args: unknown[]) => {
    console.error(`[${new Date().toISOString()}] [ERROR]`, ...args);
  },
  debug: (...args: unknown[]) => {
    console.debug(`[${new Date().toISOString()}] [DEBUG]`, ...args);
  },
};

// ============================================================================
// Task Manager
// ============================================================================

class TeamworkTaskManager {
  private config: TeamworkCronConfig;
  private activeAdwIds: Set<string>;

  constructor(config: TeamworkCronConfig) {
    this.config = config;
    this.activeAdwIds = new Set();
  }

  /**
   * Fetch eligible tasks from Teamwork project.
   */
  async getEligibleTasks(limit: number = 10): Promise<TeamworkTask[]> {
    logger.info(`Fetching tasks from Teamwork project ${this.config.project_id}`);

    const request: AgentTemplateRequest = {
      agentName: 'teamwork-task-fetcher',
      slashCommand: '/get_teamwork_tasks',
      args: [this.config.project_id, JSON.stringify(this.config.status_filter), String(limit)],
      adwId: generateShortId(),
      model: 'sonnet',
      workingDir: process.cwd(),
    };

    try {
      const response = await executeTemplate(request);

      if (!response.success) {
        logger.error(`Failed to fetch Teamwork tasks: ${response.output}`);
        return [];
      }

      // Parse JSON response
      const taskData = parseJson(response.output, z.array(z.unknown()));
      if (!taskData || taskData.length === 0) {
        logger.info('No tasks returned from Teamwork');
        return [];
      }

      // Convert to TeamworkTask objects and filter eligible
      const tasks: TeamworkTask[] = [];
      for (const taskDict of taskData) {
        try {
          const teamworkTask = TeamworkTaskSchema.parse(taskDict);

          if (isTeamworkTaskEligibleForProcessing(teamworkTask)) {
            tasks.push(teamworkTask);
            logger.info(`Found eligible task: ${teamworkTask.task_id} - ${teamworkTask.title}`);
          } else {
            logger.debug(`Skipping ineligible task: ${teamworkTask.task_id}`);
          }
        } catch (error) {
          logger.error(`Failed to parse task: ${error}`);
          continue;
        }
      }

      return tasks;
    } catch (error) {
      logger.error('Exception while fetching tasks:', error);
      return [];
    }
  }

  /**
   * Update Teamwork task status and post comment.
   */
  async updateTaskStatus(taskId: string, status: string, updateContent: string = ''): Promise<boolean> {
    logger.info(`Updating task ${taskId} to status: ${status}`);

    // Map system status to Teamwork status
    const teamworkStatus = mapStatusToTeamwork(this.config, status);

    const request: AgentTemplateRequest = {
      agentName: 'teamwork-task-updater',
      slashCommand: '/update_teamwork_task',
      args: [taskId, teamworkStatus, updateContent],
      adwId: generateShortId(),
      model: 'sonnet',
      workingDir: process.cwd(),
    };

    try {
      const response = await executeTemplate(request);

      if (response.success) {
        logger.info(`Successfully updated task ${taskId} to ${teamworkStatus}`);
        return true;
      } else {
        logger.error(`Failed to update task ${taskId}: ${response.output}`);
        return false;
      }
    } catch (error) {
      logger.error(`Exception while updating task ${taskId}:`, error);
      return false;
    }
  }

  /**
   * Delegate task to appropriate workflow script.
   */
  async delegateTask(task: TeamworkTask): Promise<boolean> {
    // Generate ADW ID
    let adwId = generateShortId();

    // Check for duplicates
    if (this.activeAdwIds.has(adwId)) {
      logger.warn(`Duplicate ADW ID detected: ${adwId}, regenerating...`);
      adwId = generateShortId();
    }

    this.activeAdwIds.add(adwId);

    // Get model preference
    const model = getTeamworkPreferredModel(task);

    // Get worktree name
    let worktreeName = task.worktree || task.tags.worktree;
    if (!worktreeName) {
      worktreeName = this.generateWorktreeName(task);
    }

    // Get task prompt
    const taskPrompt = getTeamworkTaskPromptForAgent(task);

    logger.info(`Delegating task ${task.task_id} with ADW ID ${adwId}`);
    logger.info(`  Model: ${model}`);
    logger.info(`  Worktree: ${worktreeName}`);
    logger.info(`  Workflow: ${this.determineWorkflow(task)}`);

    // Claim task immediately
    const updateMetadata = {
      adw_id: adwId,
      timestamp: new Date().toISOString(),
      model,
      worktree_name: worktreeName,
      status: 'In progress',
    };

    const claimed = await this.updateTaskStatus(task.task_id, 'In progress', JSON.stringify(updateMetadata));

    if (!claimed) {
      logger.error(`Failed to claim task ${task.task_id}`);
      this.activeAdwIds.delete(adwId);
      return false;
    }

    // Determine workflow script and arguments
    let scriptPath: string;
    let scriptArgs: string[];

    if (task.prototype) {
      scriptPath = './adws-bun/src/workflows/adw-plan-implement-update-teamwork-task.ts';
      scriptArgs = [adwId, task.task_id, taskPrompt, worktreeName, task.prototype, model, this.config.project_id];
    } else if (shouldTeamworkTaskUseFullWorkflow(task)) {
      scriptPath = './adws-bun/src/workflows/adw-plan-implement-update-teamwork-task.ts';
      scriptArgs = [adwId, task.task_id, taskPrompt, worktreeName, '', model, this.config.project_id];
    } else {
      scriptPath = './adws-bun/src/workflows/adw-build-update-teamwork-task.ts';
      scriptArgs = [adwId, task.task_id, taskPrompt, worktreeName, model, this.config.project_id];
    }

    if (this.config.dry_run) {
      logger.info(`[DRY RUN] Would spawn: ${scriptPath} ${scriptArgs.join(' ')}`);
      return true;
    }

    // Spawn detached subprocess
    try {
      const proc = Bun.spawn({
        cmd: [scriptPath, ...scriptArgs],
        env: getSafeSubprocessEnv(),
        stdout: 'ignore',
        stderr: 'ignore',
        stdin: 'ignore',
        detached: true,
      });

      proc.unref(); // Allow parent to exit without waiting

      logger.info(`Successfully spawned workflow for task ${task.task_id}`);
      return true;
    } catch (error) {
      logger.error('Failed to spawn workflow:', error);

      // Update task to Failed status
      const errorMetadata = {
        adw_id: adwId,
        error: String(error),
        timestamp: new Date().toISOString(),
      };

      await this.updateTaskStatus(task.task_id, 'Failed', JSON.stringify(errorMetadata));

      this.activeAdwIds.delete(adwId);
      return false;
    }
  }

  /**
   * Determine which workflow to use for task.
   */
  private determineWorkflow(task: TeamworkTask): string {
    if (task.prototype) {
      return `plan-implement (${task.prototype})`;
    } else if (shouldTeamworkTaskUseFullWorkflow(task)) {
      return 'plan-implement';
    } else {
      return 'build';
    }
  }

  /**
   * Generate a worktree name from task title.
   */
  private generateWorktreeName(task: TeamworkTask): string {
    let name = task.title.toLowerCase();
    // Simple sanitization
    name = name.replace(/[^a-z0-9]+/g, '-');
    name = name.replace(/^-+|-+$/g, '');
    // Limit length
    if (name.length > 50) {
      name = name.slice(0, 50);
    }
    return task.prototype ? `proto-${name}` : `feat-${name}`;
  }

  /**
   * Run a single polling cycle. Returns number of tasks processed.
   */
  async runOnce(): Promise<number> {
    logger.info('=== Starting polling cycle ===');

    // Get eligible tasks
    const tasks = await this.getEligibleTasks(this.config.max_concurrent_tasks);

    if (tasks.length === 0) {
      logger.info('No eligible tasks found');
      return 0;
    }

    logger.info(`Found ${tasks.length} eligible task(s)`);

    // Delegate each task
    let processed = 0;
    for (const task of tasks) {
      if (await this.delegateTask(task)) {
        processed++;
      }
      await Bun.sleep(1000); // Brief delay between delegations
    }

    logger.info(`=== Polling cycle complete: ${processed}/${tasks.length} tasks delegated ===`);
    return processed;
  }

  /**
   * Run continuous polling loop.
   */
  async runContinuous(): Promise<void> {
    logger.info('Starting Teamwork task monitor');
    logger.info(`Project ID: ${this.config.project_id}`);
    logger.info(`Polling interval: ${this.config.polling_interval}s`);
    logger.info(`Max concurrent tasks: ${this.config.max_concurrent_tasks}`);
    logger.info(`Status filter: ${this.config.status_filter.join(', ')}`);
    logger.info(`Dry run: ${this.config.dry_run}`);

    let cycleCount = 0;

    // Set up signal handlers for graceful shutdown
    let shouldStop = false;

    const handleSignal = (signal: string) => {
      logger.info(`\nReceived ${signal} signal, shutting down gracefully...`);
      shouldStop = true;
    };

    process.on('SIGINT', () => handleSignal('SIGINT'));
    process.on('SIGTERM', () => handleSignal('SIGTERM'));

    try {
      while (!shouldStop) {
        cycleCount++;
        logger.info(`\n--- Cycle ${cycleCount} ---`);

        await this.runOnce();

        if (shouldStop) break;

        logger.info(`Sleeping for ${this.config.polling_interval}s...`);
        await Bun.sleep(this.config.polling_interval * 1000);
      }
    } catch (error) {
      logger.error('Fatal error in polling loop:', error);
      throw error;
    }
  }
}

// ============================================================================
// Main Entry Point
// ============================================================================

async function main() {
  // Parse CLI arguments
  const { values } = parseArgs({
    options: {
      once: { type: 'boolean', default: false },
      'dry-run': { type: 'boolean', default: false },
      interval: { type: 'string' },
      'max-tasks': { type: 'string' },
    },
    allowPositionals: false,
  });

  // Load configuration from environment
  const projectId = process.env.TEAMWORK_PROJECT_ID;
  if (!projectId) {
    logger.error('TEAMWORK_PROJECT_ID environment variable is required');
    process.exit(1);
  }

  // Build configuration
  const config: TeamworkCronConfig = TeamworkCronConfigSchema.parse({
    project_id: projectId,
    polling_interval: values.interval ? parseInt(values.interval) : 15,
    max_concurrent_tasks: values['max-tasks'] ? parseInt(values['max-tasks']) : 3,
    dry_run: values['dry-run'] || false,
  });

  // Create manager
  const manager = new TeamworkTaskManager(config);

  // Run
  if (values.once) {
    logger.info('Running in single-cycle mode');
    const processed = await manager.runOnce();
    logger.info(`Processed ${processed} task(s)`);
    process.exit(0);
  } else {
    await manager.runContinuous();
  }
}

// Run if executed directly
if (import.meta.main) {
  main().catch((error) => {
    logger.error('Fatal error:', error);
    process.exit(1);
  });
}
